src/dtls.c

```diff
@@ -290,60 +354,459 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+        int cookieHashSz;
+
+        XMEMSET(&cs, 0, sizeof(cs));
+
+        /* We need to echo the session ID sent by the client */
+        if (ch->sessionId.size > ID_LEN) {
+            /* Too large. We can't echo this. */
+            ERROR_OUT(INVALID_PARAMETER, dtls13_cleanup);
+        }
+
+        /* Populate the suites struct to find a common ciphersuite */
+        XMEMSET(&suites, 0, sizeof(suites));
+        suites.suiteSz = (word16)ch->cipherSuite.size;
+        if ((suites.suiteSz % 2) != 0)
+            ERROR_OUT(INVALID_PARAMETER, dtls13_cleanup);
+        if (suites.suiteSz > WOLFSSL_MAX_SUITE_SZ)
+            ERROR_OUT(BUFFER_ERROR, dtls13_cleanup);
+        XMEMCPY(suites.suites, ch->cipherSuite.elements, suites.suiteSz);
+
+        /* Populate extensions */
+
+        /* Supported versions always need to be present. Has to appear after
+         * key share as that is the order we reconstruct it in
+         * RestartHandshakeHashWithCookie. */
+        ret = TLSX_Push(&parsedExts,
+                  TLSX_SUPPORTED_VERSIONS, ssl, ssl->heap);
+        if (ret != 0)
+            goto dtls13_cleanup;
+        /* Set that this is a response extension */
+        parsedExts->resp = 1;
+
+        ret = CopyExtensions(ssl->extensions, &parsedExts, ssl->heap);
+        if (ret != 0)
+            goto dtls13_cleanup;
+
+        /* Signature algs */
+        ret = TlsxFindByType(&tlsx, TLSX_SIGNATURE_ALGORITHMS,
+                             ch->extension);
+        if (ret != 0)
+            goto dtls13_cleanup;
+        if (tlsx.size > OPAQUE16_LEN) {
+            ato16(tlsx.elements, &len);
+            if (len != tlsx.size - OPAQUE16_LEN)
+                ERROR_OUT(BUFFER_ERROR, dtls13_cleanup);
+            if ((len % 2) != 0)
+                ERROR_OUT(BUFFER_ERROR, dtls13_cleanup);
+            suites.hashSigAlgoSz = len;
+            XMEMCPY(suites.hashSigAlgo, tlsx.elements + OPAQUE16_LEN,
+                    len);
+        }
+
+        /* Supported groups */
+        ret = TlsxFindByType(&tlsx, TLSX_SUPPORTED_GROUPS,
+                             ch->extension);
+        if (ret != 0)
+            goto dtls13_cleanup;
+        if (tlsx.size != 0) {
+            ret = TLSX_SupportedCurve_Parse(ssl, tlsx.elements,
+                                         (word16)tlsx.size, 1, &parsedExts);
+            if (ret != 0)
+                goto dtls13_cleanup;
+        }
+
+        /* Key share */
+        ret = TlsxFindByType(&tlsx, TLSX_KEY_SHARE,
+                             ch->extension);
+        if (ret != 0)
+            goto dtls13_cleanup;
+        if (tlsx.size != 0) {
+            ret = TLSX_KeyShare_Parse_ClientHello(ssl, tlsx.elements,
+                                            (word16)tlsx.size, &parsedExts);
+            if (ret != 0)
+                goto dtls13_cleanup;
+#if defined(HAVE_SESSION_TICKET) || !defined(NO_PSK)
+            haveKS = 1;
+#endif
+        }
+
```

Comment:
If the client didn't provide any `KEY_SHARE` extension (meaning no extension at all, not a 0-sized extension) we should error out in case of KE, not just provide a suggesting group in the `HRR`. `TlsxFindByType` may need some refactoring to distinguish between a 0-sized extension and a not-present one.

Reply:
Refactor to handle sent 0-len extensions. The error logic is outsourced to the functions that parse the extensions so that len checks are done in one place.